### ISSUE TITLE
[FEATURE] Produce latest output in a known location

### ISSUE THREAD

**Issue by andrey-kuprianov** [informalsystems/apalache#1077]
I have tried to upgrade [modelator](https://github.com/informalsystems/modelator) to the recent version of Apalache, but it turned out to be not as simple as I hoped for :smile:  

For some reason the Apalache output behavior has changed: now, instead of just producing an output in the current directory, as it was before, it produces it in a subdirectory like `_apalache-out/NumbersAMaxBMinTest.tla_2021-11-04T14-07-40_6241382755432900060/` , i.e. a combination of `_apalache-out`, the name of the file, plus a timestamp. The problem is that this is not printed on the output of the tool, it still prints

```
Check the counterexample in: counterexample1.tla, MC1.out, counterexample1.json
```

This makes it very difficult to interact with the tool, both for the command-line user, and automatically. E.g. if a user runs the tool multiple times, the `_apalache-out`  dir will be populated with multiple directories, and it is difficult to navigate there. Conversely, from `modelator` side, we parse the tool output, but there is no way to deduce from the output where the counterexample file is actually located. I understand that this change comes from a desire not to clog the current directory, but this just goes to the other extreme. 

Would it be possible to provide an intermediate solution? I propose the following:
 - produce the output in the directory structure as it is now
 - _additionally_, produce the _latest_ output in the directory `_apalache-out`, without subdirectories
 - print in the tool output `Check the counterexample in: _apalache-out/counterexample1.tla, _apalache-out/MC1.out, _apalache-out/counterexample1.json`

This will make it easy and clear both for the user and for other tools to deduce where the latest output is located.

P.S. I am aware of the `--out-dir` option, but this doesn't solve the problem, as the artifacts are still produced in a subdirectory with a timestamp in the name.

**Comment by shonfeder** [comment]
Just FYI We do print the output location to stdout (we should be) the line begins with `Output directory:`. I'll review this feature request tho to see how it would integrate with the current evolving design for our output management.

**Comment by andrey-kuprianov** [comment]
oh, that's nice, I have not noticed that, thanks! Still, I believe the way as proposed above would provide additional benefits:

- for the end-user: having opened the counterexample once, on subsequent calls to the tool, the counterexample will just update in the editor window automatically, compared to the necessity to open new file on each iteration
- for automatic tools a simple hard-coded relative path will work; in particular for `modelator`, no changes will be needed in the current codebase. Alternatively, deducing the output directory from the tool output will require additional logic to be implemented.

I hope those are convincing arguments.

**Comment by shonfeder** [comment]
The overall behavior and need here definitely makes sense to me. But I'm trying to think through the details a bit.

Let's assume that `_apalache-out` is analogous to a build directory, like the `_build` or `target` dirs (or similar) that many different build systems use. Whereas build systems produce "build artifacts", our output files are "checker artifacts".

In the case of build artifacts, the generated build artifacts *are* overwritten, but that's because it is assumed that the source files used are always part of the same project. So when we update the source files under the current working directory, we'll also want to update the build artifacts.

However, it seems pretty common to check a number of different specs from within the same working directory. But those the different specs you'll run the checker on don't update the counterexample output files, rather they totally change them. So overwriting the latest run results seems less obviously correct to me. Here's an example where that behavior might lead to bad UX:

- Suppose a user runs the checker on `SpecA.tla`, a counterexample is generated at `_apalache-out/counterexample1.tla`
- They open the counterexample in their editor to view it.
- They edit `SpecA.tla` and run the checker, and view the updates in a loop.
- They now switch to working on a related, but separate `SpecB.tla`, in the same working directory. They make some edits there, and run the checker.
- They now want to compare with the counter example from `SpecA.tla` so they turn to that in their editor, but actually it's been overwritten by a counterexample generated by running `SpecB.tla`!

This makes me think that we want to treat these output files a bit differently than usual build outputs in this regard. Or, viewed differently, each input TLA module produces unique set of build artifacts, and even if we want easy and stable access to one set of artifacts as they are updated, we do not want two different modules to clobber each other.

And even more obvious example of the problem here is imagining a scenario where you want to run checks on two different specs in parallel. If they both share the same files in the output directory, they'll end up clobbering each other and it'll be a mess.

I have two alternative ideas, which are not mutually exclusive:

1. Insofar as the output files are akin to build artifacts, but that in our case the source files are determined by which specs are run, not by the current working directory, how about namespacing all checker artifacts under the module names.
 
   We are currently storing each run in a directory in the out-dir named according to the pattern: `<module_name>.tla_<timestamp>.  Instead, assuming our spec is named `Counter.tla`, we could store the generated output files in subdirectories following this pattern:

   ```
   $  tree _apalache-out/Counter.tla
   _apalache-out/Counter.tla
   ├── latest
   └── runs
       ├── 2021-10-27T17-17-18_11753047800614976161
       ├── 2021-10-27T17-17-18_11753047800614976961
       └── 2021-10-27T17-17-18_11753047800614976968
```

   Wherein, `runs` will accumulate the history of different runs, and `latest` will always be updated with the most recent outputs. And we can always add more dirs into this namespace as the need arises.

2. However, there is a separate concern: these output files are channels used to communicate between programs. I think it's not nice to have to go hunting to figure out where the data your program needs is going to be output. Rather, it would be nice to just tell apalache to write the desired output data to a specific file (or, what amounts to the same, to a set of files in a dir, which is equivalent to a single file with section breaks indicated by the file name).

   To meet this need, how about we introduce an `--output` flag for the `check` subcommand that works analogously to the `--output` flag for the `parse` command?

**Comment by andrey-kuprianov** [comment]
All this seems pretty reasonable to me! `--output` flag would make it easy to integrate with automatic pipelines, while the schema for storing the model-checking artifacts also seems quite transparent for the end user. This is probably obvious, but just to make sure: I would also include into the tool output the hint to the user that the produced artifacts can be found in `_apalache-out/Counter.tla/latest`.

**Comment by konnov** [comment]
> P.S. I am aware of the --out-dir option, but this doesn't solve the problem, as the artifacts are still produced in a subdirectory with a timestamp in the name.

I think this is the main issue here. If counterexamples were written to `--out-dir`, you would not have this problem.
